File: builtins/Number.js
// Self-hosted Number prototype methods on top of __ops__
var ops = __ops__;
var ThrowTypeError = ops.ThrowTypeError;

// Save native implementations
var _nativeToFixed = Number.prototype.toFixed;
var _nativeToExponential = Number.prototype.toExponential;
var _nativeToPrecision = Number.prototype.toPrecision;

Number.isNaN = function NumberIsNaN(value) {
  return typeof value === 'number' && value !== value;
};

Number.isFinite = function NumberIsFinite(value) {
  return typeof value === 'number' && value !== Infinity && value !== -Infinity && value === value;
};

Number.isInteger = function NumberIsInteger(value) {
  return Number.isFinite(value) && Math.floor(value) === value;
};

Number.isSafeInteger = function NumberIsSafeInteger(value) {
  return Number.isInteger(value) && Math.abs(value) <= 9007199254740991;
};

// Number.prototype.toFixed (ES2025 §21.1.3.5)
Number.prototype.toFixed = function NumberToFixed(digits) {
  if (this === null || this === undefined) throw ThrowTypeError("Number.prototype.toFixed called on null or undefined");
  return _nativeToFixed.call(this, digits);
};

// Number.prototype.toExponential (ES2025 §21.1.3.4)
Number.prototype.toExponential = function NumberToExponential(digits) {
  if (this === null || this === undefined) throw ThrowTypeError("Number.prototype.toExponential called on null or undefined");
  return _nativeToExponential.call(this, digits);
};

// Number.prototype.toPrecision (ES2025 §21.1.3.6)
Number.prototype.toPrecision = function NumberToPrecision(precision) {
  if (this === null || this === undefined) throw ThrowTypeError("Number.prototype.toPrecision called on null or undefined");
  return _nativeToPrecision.call(this, precision);
};

File: crates/quench-runtime/src/builtin_meta/number.rs
//! Number method metadata.

use crate::ops::Builtin;

pub const fn fn_name(b: Builtin) -> Option<&'static str> {
    match b {
        Builtin::IsFinite => Some("Number.isFinite"),
        Builtin::IsNaN => Some("Number.isNaN"),
        Builtin::ParseFloat => Some("Number.parseFloat"),
        Builtin::ParseInt => Some("Number.parseInt"),
        Builtin::NumberIsInteger => Some("Number.isInteger"),
        Builtin::NumberIsSafeInteger => Some("Number.isSafeInteger"),
        Builtin::NumberToString => Some("Number.prototype.toString"),
        Builtin::BooleanToString => Some("Boolean.prototype.toString"),
        Builtin::NumberValueOf => Some("Number.prototype.valueOf"),
        Builtin::NumberToFixed => Some("Number.prototype.toFixed"),
        Builtin::NumberToPrecision => Some("Number.prototype.toPrecision"),
        Builtin::NumberToExponential => Some("Number.prototype.toExponential"),
        Builtin::NumberToLocaleString => Some("Number.prototype.toLocaleString"),
        _ => None,
    }
}

pub const fn fn_len(b: Builtin) -> Option<f64> {
    match b {
        Builtin::IsFinite
        | Builtin::IsNaN
        | Builtin::ParseFloat
        | Builtin::NumberIsInteger
        | Builtin::NumberIsSafeInteger => Some(1.0),
        Builtin::ParseInt => Some(2.0),
        Builtin::BooleanToString | Builtin::NumberValueOf | Builtin::NumberToLocaleString => {
            Some(0.0)
        }
        Builtin::NumberToString
        | Builtin::NumberToFixed
        | Builtin::NumberToPrecision
        | Builtin::NumberToExponential => Some(1.0),
        _ => None,
    }
}

pub const fn short_name(b: Builtin) -> Option<&'static str> {
    match b {
        Builtin::IsFinite => Some("isFinite"),
        Builtin::IsNaN => Some("isNaN"),
        Builtin::ParseFloat => Some("parseFloat"),
        Builtin::ParseInt => Some("parseInt"),
        Builtin::NumberIsInteger => Some("isInteger"),
        Builtin::NumberIsSafeInteger => Some("isSafeInteger"),
        Builtin::NumberToString | Builtin::BooleanToString => Some("toString"),
        Builtin::NumberValueOf => Some("valueOf"),
        Builtin::NumberToFixed => Some("toFixed"),
        Builtin::NumberToPrecision => Some("toPrecision"),
        Builtin::NumberToExponential => Some("toExponential"),
        Builtin::NumberToLocaleString => Some("toLocaleString"),
        _ => None,
    }
}
